File: Amazon/3900.longest-balanced-substring-after-one-swap.java
/*
 * @lc app=leetcode id=3900 lang=java
 *
 * [3900] Longest Balanced Substring After One Swap
 *
 * Prefix balance (+1 for '1', -1 for '0') gives balanced substrings when two
 * prefix values match. One swap can fix substrings whose original balance is
 * +/-2 if there is an opposite character outside; cap lengths by available
 * outside characters and binary-search prefix positions.
 *
 * Java note: HashMap<Integer,List<Integer>> groups prefix positions by balance.
 *
 * Time: O(n log n). Space: O(n).
 */

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

// @lc code=start
class Solution {
    public int longestBalanced(String s) {
        int totalZeros = 0;
        for (int i = 0; i < s.length(); i++) {
            if (s.charAt(i) == '0') {
                totalZeros++;
            }
        }
        int totalOnes = s.length() - totalZeros;

        int[] prefix = new int[s.length() + 1];
        Map<Integer, List<Integer>> positions = new HashMap<>();
        positions.computeIfAbsent(0, unused -> new ArrayList<>()).add(0);

        int balance = 0;
        for (int i = 1; i <= s.length(); i++) {
            balance += s.charAt(i - 1) == '1' ? 1 : -1;
            prefix[i] = balance;
            positions.computeIfAbsent(balance, unused -> new ArrayList<>()).add(i);
        }

        int answer = 0;
        Map<Integer, Integer> earliest = new HashMap<>();
        for (int i = 0; i < prefix.length; i++) {
            Integer first = earliest.get(prefix[i]);
            if (first == null) {
                earliest.put(prefix[i], i);
            } else {
                answer = Math.max(answer, i - first);
            }
        }

        int capTooManyOnes = 2 * totalZeros;
        int capTooManyZeros = 2 * totalOnes;
        for (int right = 1; right < prefix.length; right++) {
            int bal = prefix[right];
            answer = Math.max(answer, bestWithCap(positions.getOrDefault(bal - 2, List.of()), right, capTooManyOnes));
            answer = Math.max(answer, bestWithCap(positions.getOrDefault(bal + 2, List.of()), right, capTooManyZeros));
        }
        return answer;
    }

    private int bestWithCap(List<Integer> starts, int right, int cap) {
        if (cap <= 0) {
            return 0;
        }
        int index = lowerBound(starts, right - cap);
        if (index < starts.size() && starts.get(index) < right) {
            return right - starts.get(index);
        }
        return 0;
    }

    private int lowerBound(List<Integer> values, int target) {
        int lo = 0;
        int hi = values.size();
        while (lo < hi) {
            int mid = (lo + hi) >>> 1;
            if (values.get(mid) < target) {
                lo = mid + 1;
            } else {
                hi = mid;
            }
        }
        return lo;
    }
}
// @lc code=end

File: companies/Amazon/3900.longest-balanced-substring-after-one-swap.go
package leetcode

//
// @lc app=leetcode id=3900 lang=golang
//
// [3900] Longest Balanced Substring After One Swap
//
// Notes
// Prefix balance (+1 for '1', -1 for '0') gives the longest already-balanced
// substring by equal prefix balances. One swap can fix substrings whose balance
// differs by 2, subject to having an opposite character outside; prefix-position
// lists plus binary search find the earliest allowed start under the length cap.
// Time: O(n log n). Space: O(n).
//
// @lc code=start

import "sort"

func LongestBalanced3900(s string) int {
	totalZeros := 0
	for i := range s {
		if s[i] == '0' {
			totalZeros++
		}
	}
	totalOnes := len(s) - totalZeros

	prefix := []int{0}
	positions := map[int][]int{0: {0}}
	balance := 0
	for index := 1; index <= len(s); index++ {
		if s[index-1] == '1' {
			balance++
		} else {
			balance--
		}
		prefix = append(prefix, balance)
		positions[balance] = append(positions[balance], index)
	}

	answer := 0
	earliest := map[int]int{}
	for index, bal := range prefix {
		if first, ok := earliest[bal]; ok {
			if index-first > answer {
				answer = index - first
			}
		} else {
			earliest[bal] = index
		}
	}

	capTooManyOnes := 2 * totalZeros
	capTooManyZeros := 2 * totalOnes
	for right, bal := range prefix {
		if right == 0 {
			continue
		}
		if best := bestWithCap3900(positions[bal-2], right, capTooManyOnes); best > answer {
			answer = best
		}
		if best := bestWithCap3900(positions[bal+2], right, capTooManyZeros); best > answer {
			answer = best
		}
	}

	return answer
}

func bestWithCap3900(starts []int, right int, cap int) int {
	if cap <= 0 {
		return 0
	}
	index := sort.SearchInts(starts, right-cap)
	if index < len(starts) && starts[index] < right {
		return right - starts[index]
	}
	return 0
}

// @lc code=end
